src/pointer.rs

```diff
@@ -99,32 +42,50 @@ const fn is_valid_ptr(value: &str) -> bool {
 #[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
 pub struct Pointer(str);
 
+impl Default for &'static Pointer {
+    fn default() -> Self {
+        Pointer::root()
+    }
+}
 impl core::fmt::Display for Pointer {
     fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
         self.0.fmt(f)
     }
 }
-
 impl Pointer {
     /// Private constructor for strings that are known to be correctly encoded
     ///
     /// This is a zero-copy constructor
     fn new<S: AsRef<str> + ?Sized>(s: &S) -> &Self {
-        unsafe { &*(s.as_ref() as *const str as *const Self) }
+        unsafe { &*(core::ptr::from_ref::<str>(s.as_ref()) as *const Self) }
```

Comment:
Out of interest, was this clippy's suggestion? I originally copied the expression from `std::path::Path` essentially verbatim, so I'm amused clippy wasn't happy with it :p

Reply:
It was a clippy lint, ya. I wasn't sure about if I should apply the suggestion. I know very little working in unsafe. I probably should have avoided it. 

Sorry about that.

Reply:
No, no, no. This is perfectly fine, I was just curious. I'd expect std to apply clippy lints themselves, but maybe it's a new lint or something.

Reply:
Also, I should probably have left a link to the reference implementation: https://doc.rust-lang.org/src/std/path.rs.html#1-3385

I'm no unsafe expert either, but I think this one is straightforward. It's just transmuting from a reference to an `str` to a reference to a `Pointer`, without copying, which relies on the internal representation of either being the same. Writing this down just now made me realise I missed one important detail in the definition of `Pointer`: `#[repr(transparent)]`. Will push a small PR to fix that.

Reply:
https://github.com/chanced/jsonptr/pull/49

Reply:
Sorry, I was wrapping up some errands earlier.

This is the lint that was tripping: https://rust-lang.github.io/rust-clippy/master/index.html#ref_as_ptr

> # ref_as_pt
> What it does
>
> Checks for casts of references to pointer using as and suggests `std::ptr::from_ref` and `std::ptr::from_mut` instead.
> 
> Why is this bad?
> Using as casts may result in silently changing mutability or type.
>
> Example
> ```rust
> let a_ref = &1;
> let a_ptr = a_ref as *const _;
> ```
> Use instead:
> ```rust
> let a_ref = &1;
> let a_ptr = std::ptr::from_ref(a_ref);
>

Reply:
Added in 1.78.0. Yep, pretty new. Thanks.